lib/asyn-gai.c

```diff
@@ -0,0 +1,493 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) 1998 - 2021, Daniel Stenberg, <daniel@haxx.se>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#if USE_AGRES
+
+#include <netdb.h>
+#ifdef HAVE_SIGNAL_H
+# include <signal.h>
+#endif
+#ifdef HAVE_PTHREAD_H
+# include <pthread.h>
+#endif
+
+#include "urldata.h"
+#include "sendf.h"
+#include "hostip.h"
+#include "hash.h"
+#include "share.h"
+#include "strerror.h"
+#include "url.h"
+#include "multiif.h"
+#include "inet_ntop.h"
+#include "curl_threads.h"
+#include "connect.h"
+/* The last 3 #include files should be in this order */
+#include "curl_printf.h"
+#include "curl_memory.h"
+#include "memdebug.h"
+
+struct resolver {
+  struct {
+    curl_mutex_t lock;
+    int flag;
+  } done;
+  struct {
+    struct curltime start;
+    timediff_t left;
+    timediff_t expire;
+  } tm;
+  struct {
+    struct sigevent event;
+    struct addrinfo hint;
+    struct gaicb cb_buf;
+    struct gaicb *cb_ptr[1];
+  } gai;
+  struct {
+    char *hostname;
+    char service[12];
+  } data;
+};
+
+#define AGRES_DEBUG 0
+#define AGRES_EXPIRE_INIT 2U
+#define AGRES_EXPIRE_GROW 2U
+#define AGRES_EXPIRE_MAX (2U<<8)
+
+static inline void set_done(struct resolver *res)
+{
+  DEBUGASSERT(res);
+  if(res) {
+    Curl_mutex_acquire(&res->done.lock);
+    res->done.flag = 1;
+    Curl_mutex_release(&res->done.lock);
+  }
+}
+
+static inline int done(struct resolver *res)
```

Comment:
`inline` is not a C89 keyword I think you can drop that and let the compiler decide this for itself